_posts/2019-06-05-uncertainty-estimates-tfprobability/uncertainty_tfprobability.Rmd

```diff
@@ -0,0 +1,782 @@
+author:
+  - name: Sigrid Keydana
+    affiliation: RStudio
+    affiliation_url: https://www.rstudio.com/
+slug: keydana2019uncertaintytfp
+date: 06-05-2019
+categories:
+  - Probability and statistics 
+  - Keras
+  - Concepts
+output:
+  distill::distill_article:
+    self_contained: false
+preview: images/uci_both.png
+---
+
+
+```{r setup, include=FALSE}
+knitr::opts_chunk$set(echo = TRUE, eval = FALSE)
+```
+
+
+About six months ago, we showed [how to create a custom wrapper](https://blogs.rstudio.com/tensorflow/posts/2018-11-12-uncertainty_estimates_dropout/) to obtain uncertainty estimates from a Keras network. Today we present a less laborious, as well faster-running, way using [tfprobability](https://rstudio.github.io/tfprobability/), the R wrapper to TensorFlow Probability. Like most posts on this blog, this one won't be short, so let's quickly state what you can expect in return of reading time.
+
+## What to expect from this post
+
+Starting from what _not_ to expect: There won't be a recipe that tells you how _exactly_ to set all parameters involved in order to report the "right" uncertainty measures. But then, what _are_ the "right" uncertainty measures? Unless you happen to work with a method that has no (hyper-)parameters to tweak, there will always be uncertainty about how to report uncertainty.
+
+What you _can_ expect though is an introduction to obtaining uncertainty estimates for Keras networks, as well as an empirical report of how tweaking (hyper-)parameters may affect the results. As in the aforementioned post, we perform our tests on a simulated as well as on a real dataset, the [Combined Cycle Power Plant Data Set](http://archive.ics.uci.edu/ml/datasets/Combined+Cycle+Power+Plant). At the end, in place of strict rules you should have acquired some intuition that will transfer to other real-world datasets.
+
+Did you notice our talking about _Keras networks_ above? Indeed this post has an additional goal: So far, we haven't really discussed yet how `tfprobability` goes together with `keras`. Now we finally do (in short: they work together seemlessly).
+
+Finally, the notions of _aleatoric_ and _epistemic_ uncertainty, which may have stayed a bit abstract in the prior post, should get much more concrete here.
+
+## Aleatoric vs. epistemic uncertainty
+
+Reminiscent somehow of the classic decomposition of generalization error into bias and variance, splitting uncertainty into its epistemic and aleatoric constituents separates an irreducible from a reducible part.
```

Comment:
If there's a piece somewhere that you like on decomposition of generalization error, it might be nice to have a recommended reading/link here. If not, it's not a big deal!

Reply:
Also, maybe add hover-able definitions for epistemic and aleatoric? I've run into quite a few folks unfamiliar with the latter term (I knew it from aleatoric music — good old John Cage's 4'33 of silence, for one). I don't think you need anything big, but it's nice to throw the reader a bone, and keep them from having to look it up!

Reply:
You, in effect, define them below, but I also think they're just cool words!